File: MPC/Optimizer.py
import numpy as np
from MPC.Plant import *
# from time import sleep

def fdm2c(g, x, h=1e-6):
    # initialize parameters
    Nx = len( x );        # dimension of input
    Ng = len( g(x) );     # dimension of output

    # calculate derivative at each input
    grad = np.empty( (Nx,Ng) );
    for i in range( Nx ):
        xn1 = x.copy();  xn1[i] = xn1[i] - h;
        xp1 = x.copy();  xp1[i] = xp1[i] + h;

        gn1 = g( xn1 );
        gp1 = g( xp1 );

        grad[i,:] = (gp1 - gn1).reshape(Ng,)/(2*h);

    return grad;

class Cost:
    def __init__(self, g,
            gradient=None, hessian=None):
        # set cost function
        self.cost = g;

        # set gradient function
        if gradient is None:
            self.grad = lambda x: fdm2c( self.cost, x );
        else:
            self.grad = gradient;

        # set hessian function
        if hessian is None:
            self.hess = lambda x: fdm2c( self.grad, x );
        else:
            self.hess = hessian;

# Class: Optimizer()
# Assumptions:
#   1). Objective terms can be/are vectorized.
#   2). Cost function is scalar with a single state input term.
#       c = g(x)
class Optimizer( Cost ):
    def __init__(self, g, eps=1e-6, solver='ngd'):
        # inherit cost function class parameters
        Cost.__init__( self, g );

        # tolerance for zero and gradient descent step-size
        self.max_iter = 10;
        self.eps = eps;  # zero-approximation
        self.alpha = 0.1;

        # set solution step function
        self.setStepMethod( solver );

    def setStepSize(self, a):
        self.alpha = a;
        # Return instance of self.
        return self;

    def setMaxIter(self, n):
        self.max_iter = n;
        # Return instance of self.
        return self;

    def setObjectiveFunction(self, g):
        self.cost = g;
        # Return instance of self.
        return self;

    def setStepMethod(self, solver):
        self.solver = solver;
        if self.solver == 'ngd':
            self.step = lambda x, dg: x - self.alpha*dg;
        if self.solver == 'nno':
            self.step = None;  # not setup

        # Return instance of self.
        return self;

    def solve(self, xinit, verbose=0, shape=None):
        x = xinit.copy();  # copy the initial guess
        dg = self.grad( x );
        gnorm = np.linalg.norm( dg );

        n = 1;
        while gnorm > self.eps:
            # Calculate new state, gradient and norm.
            x = self.step( x, dg );
            dg = self.grad( x );
            gnorm = np.linalg.norm( dg );

            # If solution in verbose mode.
            if verbose:
                # print("Gradient:  ", g);
                print( "\n|g|:    \t", gnorm );
                print( "New Cost: \t", self.cost( x )[0] );
                print( "New Input:\n", x.reshape( shape ).T );

            # If max number of iterations reached.
            if n > self.max_iter:
                break;
            n += 1;

        # Return solution.
        return x.reshape( shape );

# Class: ModelPredictiveControl()
# Assumptions:
#   1). Instantaneous cost is a function of the state AND input:
#       c = g(x,u)
#   2). Period cost = Terminal cost, and
#   3). Initial state does not change during, or have impact on opt.:
#       G(X,U) = G(X[1:],U)
class ModelPredictiveControl( Model, Optimizer ):
    def __init__(self, F, g,
            P=10, k=1, Nx=1, Nu=1, dt=1e-3,
            model_type='discrete',
            cost_type='instant'):
        # Save model to inherited class.
        Model.__init__( self, F,
                dt=dt, model_type=model_type );

        # Initialize optimizer with g=None.
        self.cost_type = cost_type;
        self.givencost = g;
        Optimizer.__init__(self, None);  # fragile

        # Dimensions parameters.
        self.P = P;
        self.k = k;
        self.Nu = Nu;
        self.Nx = Nx;
        self.alpha = 0.1;

    def statePrediction(self, x0, uList):
        # Intialize empty state set.
        xList = np.empty( (self.Nx, self.P+1) );
        xList[:,0] = x0[:,0];

        # Propagate model using input set for length of horizon.
        for i in range( self.P ):
            u = uList[:,i,None];
            x = xList[:,i,None];

            # For a given number of 'knots' hold the input constant.
            for j in range( self.k ):
                x = self.prop( x, u );

            # Add to set.
            xList[:,i+1] = x[:,0];

        # Return state prediction set.
        return xList;

    def costPrediction(self, x0, uList):
        # Generate the prediction states.
        xList = self.statePrediction( x0, uList );

        # If the cost is instantaneous...
        if self.cost_type == 'instant':
            C = [0];
            for i in range( self.P ):
                u = uList[:,i,None];
                x = xList[:,i+1,None];
                C += self.givencost( x, u );

        # If the cost is horizon...
        elif self.cost_type == 'horizon':
            C = self.givencost( xList, uList );

        # Return the calculated cost.
        return C

    def costFunctionGenerator(self, x0):
        # Generate cost function around x0.
        return lambda uList: \
            self.costPrediction( x0, uList.reshape( self.Nu,self.P ) );

    def solve(self, x0, uinit, verbose=0):
        # If necessary set Nx to number of states.
        if self.Nx is None:
            self.Nx = len( x0 );

        # Initialize optimization variable with cost generator.
        self.setObjectiveFunction( self.costFunctionGenerator( x0 ) );

        # Find optimization results and return.
        uvect = uinit.reshape( self.Nu*self.P, 1 );
        ufinal = Optimizer.solve( self, uvect,
            verbose=verbose, shape=(self.Nu,self.P) );
        return ufinal;

File: MPC/cpp/Optimizer.cpp

#include "Helpers.h"
#include "Optimizer.h"

namespace nap
{
// Class: Cost()
    Cost::Cost(MatrixXd (*g)(const MatrixXd &)):
        Cost(g, 1000) {}

    Cost::Cost(MatrixXd (*g)(const MatrixXd &), const int &n):
        Cost(g, n, 0.1) {}

    Cost::Cost(MatrixXd (*g)(const MatrixXd &), const int &n, const double &a):
        Cost(g, n, a, 1e-6) {}

    Cost::Cost(MatrixXd (*g)(const MatrixXd &), const int &n, const double &a, const double &h):
        Cost(g, n, a, h, 1e-6) {}

    Cost::Cost(MatrixXd (*g)(const MatrixXd &), const int &n, const double &a, const double &h, const double &e):
        cost(g), max_iter(n), alpha(a), step_size(h), epsilon(e) {}

    MatrixXd Cost::gradient(const MatrixXd &x)
    {
        return fdm2c( cost, x, step_size );
    }

    MatrixXd Cost::step(const MatrixXd &x, const MatrixXd &dg)
    {
        return x - alpha*dg;
    }

    MatrixXd Cost::solve(const MatrixXd &xinit)
    {
        // Initialize iteration variable at x0.
        MatrixXd x = xinit;
        MatrixXd dg = gradient( x );

        // Optimization loop.
        int n = 0;
        while (dg.norm() > epsilon)
        {
            x = step( x, dg );
            dg = gradient( x );

            n += 1;
            if (n == max_iter) {
                std::cout << "WARNING: Iteration break executed." << std::endl;
                break;
            }
        }

        // Return minimum of cost function.
        return x;
    }

// Class: ModelCost()
    ModelCost::ModelCost(const Plant &f, const Cost &gx, const Cost &gu):
        mvar(f), costx(gx), costu(gu) {}

    ModelCost::ModelCost(MatrixXd (*f)(const MatrixXd &, const MatrixXd &), MatrixXd (*gx)(const MatrixXd &), MatrixXd (*gu)(const MatrixXd &)):
        mvar(f), costx(gx), costu(gu) {}

    MatrixXd ModelCost::cost(const MatrixXd &x, const MatrixXd &u)
    {
        return costx.cost(x) + costu.cost(u);
    }

    MatrixXd ModelCost::xgradient(const MatrixXd &x)
    {
        return fdm2c(costx.cost, x, costx.step_size);
    }

    MatrixXd ModelCost::ugradient(const MatrixXd &u)
    {
        return fdm2c(costu.cost, u, costu.step_size);
    }

    MatrixXd ModelCost::gradient(const MatrixXd &x, const MatrixXd &u)
    {
        return xgradient( x ) + ugradient( u );
    }

// Class: PredictiveCost()
    PredictiveCost::PredictiveCost(const Plant &f, const Cost &gx, const Cost &gu):
        PredictiveCost(f, gx, gu, 10, 1) {}

    PredictiveCost::PredictiveCost(const Plant &f, const Cost &gx, const Cost &gu, const int &P):
        PredictiveCost(f, gx, gx, P, 1) {}

    PredictiveCost::PredictiveCost(const Plant &f, const Cost &gx, const Cost &gu, const int &P, const int &k):
        ModelCost(f, gx, gx),
        horz_length(P),
        knot_length(k) {}

    MatrixXd PredictiveCost::prediction(const MatrixXd &xinit, const MatrixXd &ulist)
    {
        // Dimensions of simulation.
        const int N = xinit.rows();

        // Check that ulist is properly dimensioned.
        if (ulist.cols() != horz_length) {
            std::cout << "ERROR: 'ulist' is not properly dimensioned." << std::endl;
            return MatrixXd::Zero(N, horz_length+1);
        }

        // Simulation set and initial conditions.
        MatrixXd xlist(N,horz_length+1);
        xlist.col(0) = xinit;

        // Simulation loop.
        for (int i(0); i < horz_length; ++i) {
            xlist.col(i+1) = mvar.prop(xlist.col(i), ulist.col(i));
        }

        // Return simulation set.
        return xlist;
    }

    MatrixXd PredictiveCost::cost(const MatrixXd &xinit, const MatrixXd &ulist)
    {
        // Check that ulist is properly dimensioned.
        MatrixXd C(1,1);  C << -1;
        if (ulist.cols() != horz_length) {
            std::cout << "ERROR: 'ulist' is not properly dimensioned." << std::endl;
            return C;
        }

        // Get dimension and simulation set.
        const int N = xinit.rows();
        MatrixXd xlist = prediction(xinit, ulist);

        // Iterate through list summing cost.
        C << 0;
        for (int i(0); i < horz_length; ++i) {
            C += cost(xlist.col(i), ulist.col(i));
        }

        // Cost of final state (assumed zero input).
        C += cost(xlist.col(horz_length), MatrixXd::Zero(N,1));

        // Return cumulative cost.
        return C;
    }

    // TODO: Function is messy. I would like it to be more generalized.
    MatrixXd PredictiveCost::ugradient(const MatrixXd &xinit, const MatrixXd &ulist)
    {
        // Initialization of gradient vector.
        const int M = ulist.rows();

        // Manually calculate gradient vector.
        MatrixXd up(M,horz_length);  MatrixXd Cp(M,horz_length);
        MatrixXd un(M,horz_length);  MatrixXd Cn(M,horz_length);
        for (int i(0); i < M; ++i) {
            for (int j(0); j < horz_length; ++j) {
                up = ulist;  up(i,j) = ulist(i,j) + costu.step_size;
                un = ulist;  un(i,j) = ulist(i,j) - costu.step_size;

                Cp(i,j) = cost(xinit, up)(0,0);
                Cn(i,j) = cost(xinit, un)(0,0);
            }
        }

        // Return gradient (as matrix).
        return (Cp - Cn)/(2*costu.step_size);
    }
}
